### PR TITLE
Thread.alive?

### PR DESCRIPTION
**Pull request by bararchy**
Added specs
Added alive? to thread.cr
Added try_kill to libpthread

### CODE REVIEW

File: src/thread/thread.cr [hunk 1]
```diff
@@ -22,6 +22,14 @@ class Thread(T, R)
     LibPThread.detach(@th) unless @detached
   end
 
+  def alive?
+    if LibPThread.try_kill(@th, 0) == 0
```
Comment (jhass) [inline src/thread/thread.cr]: Well, it's not a try, it's the `0` that makes it a noop, I'd prefer keep calling it kill though.

File: src/thread/thread.cr [hunk 2]
```diff
@@ -22,6 +22,14 @@ class Thread(T, R)
...
+      return true
+    else
+      return false
+    end
+  end
```
Comment (jhass) [inline src/thread/thread.cr]: No  need the if else, `==` already returns a `Bool`.

File: spec/std/thread_spec.cr [hunk 3]
```diff
@@ -8,6 +8,13 @@ describe "Thread" do
     a.should eq(1)
   end
 
+  it "checks if thread is alive" do 
+    thread = Thread.new { a = 5}
+    thread.alive?.should eq(true)
+    thread.join
+    thread.alive?.should eq(false)
+  end
```
Comment (jhass) [inline spec/std/thread_spec.cr]: We got nice aliases here :) http://docs.travis-ci.com/user/firefox/

**Comment by ysbaddaden** [comment]
This is unsafe as per the `pthread_kill` manpage:

> POSIX.1-2008 recommends that if an implementation detects the use of a thread ID after the end of its lifetime, pthread_kill() should return the error ESRCH. The glibc implementation returns this error in the cases where an invalid thread ID can be detected. But note also that POSIX says that an attempt to use a thread ID whose lifetime has ended produces undefined behavior, and an attempt to use an invalid thread ID in a call to pthread_kill() can, for example, cause a segmentation fault.

**Comment by jhass** [comment]
Darn, looks like MRI maintains its own flag: http://rxr.whitequark.org/mri/source/thread.c#2575

**Comment by bararchy** [comment]
The only other option I saw was to use "pthread_tryjoin_np"

But it gave me some segfaults as well, but, this also could be because I didn't use it well

**Comment by bararchy** [comment]
This might be stupid, but, what about doing something like
@jhass @ysbaddaden 
Can the pointer be a Thread even if the Thread died ? (as in joined?) or would he be deleted by the GC ?

```crystal
t = Thread.new { a = 5}
check =  pointerof(t)
puts "alive" if check.is_a?(Pointer(Thread))  
puts "dead" if !check.is_a?(Pointer(Thread))
```

**Comment by jhass** [comment]
A dead thread is still a thread, it's just not running anymore.

**Comment by bararchy** [comment]
Can we add this to Thread and put a note saying that a use of this option is for now "unsafe" ? 
I did more the 100000 runs of this check, not even once did it segfaulted.

**Comment by ysbaddaden** [comment]
It may pass on a Linux glibc, but be unsafe in other versions, libc or platforms, and complicate ports. The problem is the undefined behavior: https://en.wikipedia.org/wiki/Undefined_behavior

Note: you don't need to push it into crystal to use it.

**Comment by asterite** [comment]
Note that in the long run to do concurrent things you will have to use spawn and channels. Threads will be used by the runtime scheduler to manager the coroutines/fibers, so it doesn't matter much for the end-user what Thread will end up having or looking like (thought we might need something to know if a Thread is alive). Maybe Thread will be there, maybe with another name, but it's usage will be discouraged.

**Comment by bararchy** [comment]
@asterite Using spawn is almost like using Fiber.new right ? 
Doesn't it means that a code invoked with spawn won't have the same performance like Thread.new ? as 2 Fibers cannot really be running concurrently right ? won't that just take us to GIL issue like Ruby, but just without the Interpretor part.

**Comment by asterite** [comment]
As we don't expose Thread to the user, this shouldn't be needed. If later we need it internally for the scheduler, we'll probably do it with a flag inside the thread or something like that.

**Pull request closed without merging**